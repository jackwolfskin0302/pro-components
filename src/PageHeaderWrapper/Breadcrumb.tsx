import React from 'react';
import pathToRegexp from 'path-to-regexp';
import { urlToList } from '../utils/pathTools';
import { MenuDataItem, MessageDescriptor } from '../typings';
import { BreadcrumbProps as AntdBreadcrumbProps } from 'antd/lib/breadcrumb';
import * as H from 'history';
import { Settings } from '../defaultSettings';

export interface BreadcrumbProps {
  breadcrumbList?: Array<{ title: string; href: string }>;
  home?: string;
  location?: H.Location;
  menu?: Settings['menu'];
  breadcrumb?: { [path: string]: MenuDataItem };
  formatMessage?: (message: MessageDescriptor) => string;
}

// 渲染Breadcrumb 子节点
// Render the Breadcrumb child node
const itemRender: AntdBreadcrumbProps['itemRender'] = (
  route,
  params,
  routes,
  paths,
) => {
  const last = routes.indexOf(route) === routes.length - 1;
  return last || !route.component ? (
    <span>{route.breadcrumbName}</span>
  ) : (
    <a>{route.breadcrumbName}</a>
  );
};

const renderItemLocal = (
  item: MenuDataItem,
  props: BreadcrumbProps,
): string => {
  const {
    formatMessage,
    menu = {
      locale: false,
    },
  } = props;
  if (item.locale && formatMessage) {
    if (menu.locale) {
      return formatMessage({ id: item.locale });
    }
  }
  return item.name as string;
};

export const getBreadcrumb = (
  breadcrumb: { [path: string]: MenuDataItem },
  url: string,
): MenuDataItem => {
  if (!breadcrumb) {
    return {
      path: '',
    };
  }
  let breadcrumbItem = breadcrumb[url];
  if (!breadcrumbItem) {
    Object.keys(breadcrumb).forEach(item => {
      if (pathToRegexp(item).test(url)) {
        breadcrumbItem = breadcrumb[item];
      }
    });
  }
  return breadcrumb || { path: '' };
};

export const getBreadcrumbProps = (props: BreadcrumbProps) => {
  const { location, breadcrumb } = props;
  return {
    location,
    breadcrumb,
  };
};

// Generated according to props
const conversionFromProps = (
  props: BreadcrumbProps,
): AntdBreadcrumbProps['routes'] => {
  const { breadcrumbList = [] } = props;
  return breadcrumbList
    .map(item => {
      const { title, href } = item;
      return {
        path: href,
        breadcrumbName: title,
      };
    })
    .filter(item => item.path);
};

const conversionFromLocation = (
  routerLocation: BreadcrumbProps['location'],
  breadcrumb: { [path: string]: MenuDataItem },
  props: BreadcrumbProps,
): AntdBreadcrumbProps['routes'] => {
  if (!routerLocation) {
    return [];
  }
  const { home } = props;
  // Convert the url to an array
  const pathSnippets = urlToList(routerLocation.pathname);
  // Loop data mosaic routing
  const extraBreadcrumbItems: AntdBreadcrumbProps['routes'] = pathSnippets
    .map(url => {
      const currentBreadcrumb = getBreadcrumb(breadcrumb, url);
      if (currentBreadcrumb.inherited) {
        return { path: '', breadcrumbName: '' };
      }
      const name = renderItemLocal(currentBreadcrumb, props);
      const { hideInBreadcrumb } = currentBreadcrumb;
      return name && !hideInBreadcrumb
        ? {
            path: url,
            breadcrumbName: name,
          }
        : { path: '', breadcrumbName: '' };
    })
    .filter(item => item && item.path);
  // Add home breadcrumbs to your head if defined
  if (home) {
    extraBreadcrumbItems.unshift({
      path: '/',
      breadcrumbName: home,
    });
  }
  return extraBreadcrumbItems;
};

export type BreadcrumbListReturn = Pick<
  AntdBreadcrumbProps,
  Extract<keyof AntdBreadcrumbProps, 'routes' | 'itemRender'>
>;

/**
 * 将参数转化为面包屑
 * Convert parameters into breadcrumbs
 */
export const conversionBreadcrumbList = (
  props: BreadcrumbProps,
): BreadcrumbListReturn => {
  const { breadcrumbList } = props;
  const { location, breadcrumb } = getBreadcrumbProps(props);
  if (breadcrumbList && breadcrumbList.length) {
    return {
      routes: conversionFromProps(props),
      itemRender,
    };
  }

  // 根据 location 生成 面包屑
  // Generate breadcrumbs based on location
  if (location && location.pathname && breadcrumb) {
    return {
      routes: conversionFromLocation(location, breadcrumb, props),
      itemRender,
    };
  }
  return {
    routes: [],
  };
};
