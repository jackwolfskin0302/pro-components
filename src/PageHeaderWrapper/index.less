@import '~antd/lib/style/themes/default.less';

@ant-pro-page-header-wrap: ~'@{ant-prefix}-pro-page-header-wrap';

.@{ant-pro-setting-drawer}-children-content {
  margin: 24px 24px 0;
}
.@{ant-pro-setting-drawer}-page-header-warp {
  background-color: @component-background;
}

.@{ant-pro-setting-drawer}-main {
  .@{ant-pro-setting-drawer}-detail {
    display: flex;
  }

  .@{ant-pro-setting-drawer}-row {
    display: flex;
    width: 100%;
  }

  .@{ant-pro-setting-drawer}-title-content {
    margin-bottom: 16px;
  }

  @media screen and (max-width: @screen-sm) {
    .@{ant-pro-setting-drawer}-content {
      margin: 24px 0 0;
    }
  }

  .@{ant-pro-setting-drawer}-title,
  .@{ant-pro-setting-drawer}-content {
    flex: auto;
  }

  .@{ant-pro-setting-drawer}-extraContent,
  .@{ant-pro-setting-drawer}-main {
    flex: 0 1 auto;
  }

  .@{ant-pro-setting-drawer}-main {
    width: 100%;
  }

  .@{ant-pro-setting-drawer}-title {
    margin-bottom: 16px;
  }

  .@{ant-pro-setting-drawer}-logo,
  .@{ant-pro-setting-drawer}-content,
  .@{ant-pro-setting-drawer}-extraContent {
    margin-bottom: 16px;
  }

  .@{ant-pro-setting-drawer}-extraContent {
    min-width: 242px;
    margin-left: 88px;
    text-align: right;
  }
}

@media screen and (max-width: @screen-xl) {
  .@{ant-pro-setting-drawer}-extraContent {
    margin-left: 44px;
  }
}

@media screen and (max-width: @screen-lg) {
  .@{ant-pro-setting-drawer}-extraContent {
    margin-left: 20px;
  }
}

@media screen and (max-width: @screen-md) {
  .@{ant-pro-setting-drawer}-row {
    display: block;
  }

  .@{ant-pro-setting-drawer}-action,
  .@{ant-pro-setting-drawer}-extraContent {
    margin-left: 0;
    text-align: left;
  }
}

@media screen and (max-width: @screen-sm) {
  .@{ant-pro-setting-drawer}-detail {
    display: block;
  }
}
