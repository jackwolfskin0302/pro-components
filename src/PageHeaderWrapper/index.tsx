import { PageHeader, Tabs, Typography } from 'antd';
import React from 'react';
import { TabsProps } from 'antd/lib/tabs';
import './index.less';
import GridContent from '../GridContent';
import RouteContext from '../RouteContext';

interface PageHeaderTabConfig {
  tabList?: {
    key: string;
    tab: string;
  }[];
  tabActiveKey?: TabsProps['activeKey'];
  onTabChange?: TabsProps['onChange'];
  tabBarExtraContent?: TabsProps['tabBarExtraContent'];
}

interface PageHeaderWrapperProps extends PageHeaderTabConfig {
  content?: React.ReactNode;
  title?: React.ReactNode;
  extraContent?: React.ReactNode;
}

const prefixedClassName = 'ant-pro-page-header-wrap';

/**
 * render Footer tabList
 * In order to be compatible with the old version of the PageHeader
 * basically all the functions are implemented.
 */
const renderFooter: React.SFC<PageHeaderWrapperProps> = ({
  tabList,
  tabActiveKey,
  onTabChange,
  tabBarExtraContent,
}) => {
  return tabList && tabList.length ? (
    <Tabs
      className={`${prefixedClassName}-tabs`}
      activeKey={tabActiveKey}
      onChange={key => {
        if (onTabChange) {
          onTabChange(key);
        }
      }}
      tabBarExtraContent={tabBarExtraContent}
    >
      {tabList.map(item => (
        <Tabs.TabPane tab={item.tab} key={item.key} />
      ))}
    </Tabs>
  ) : null;
};

const renderPageHeader = (
  content: React.ReactNode,
  extraContent: React.ReactNode,
): React.ReactNode => {
  if (!content && !extraContent) {
    return null;
  }
  return (
    <div className={`${prefixedClassName}-detail`}>
      <div className={`${prefixedClassName}-main`}>
        <div className={`${prefixedClassName}-row`}>
          {content && (
            <div className={`${prefixedClassName}-content`}>{content}</div>
          )}
          {extraContent && (
            <div className={`${prefixedClassName}-extraContent`}>
              {extraContent}
            </div>
          )}
        </div>
      </div>
    </div>
  );
};

const PageHeaderWrapper: React.SFC<PageHeaderWrapperProps> = ({
  children,
  title,
  content,
  extraContent,
  ...restProps
}) => (
  <RouteContext.Consumer>
    {value => (
      <div style={{ margin: '-24px -24px 0' }}>
        <div className={`${prefixedClassName}-page-header-warp`}>
          <GridContent>
            <PageHeader
              {...value}
              title={
                <Typography.Title
                  level={4}
                  style={{
                    margin: 0,
                  }}
                >
                  {title || value.title}
                </Typography.Title>
              }
              {...restProps}
              footer={renderFooter(restProps)}
            >
              {renderPageHeader(content, extraContent)}
            </PageHeader>
          </GridContent>
        </div>
        {children ? (
          <GridContent>
            <div className={`${prefixedClassName}-children-content`}>
              {children}
            </div>
          </GridContent>
        ) : null}
      </div>
    )}
  </RouteContext.Consumer>
);

export default PageHeaderWrapper;
