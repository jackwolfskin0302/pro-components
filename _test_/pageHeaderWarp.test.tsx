import { render } from 'enzyme';

import React from 'react';
import PageHeaderWrapper from '../src/PageHeaderWrapper';

describe('BasicLayout', () => {
  it('base use', () => {
    const html = render(<PageHeaderWrapper />).html();
    expect(html).toMatchSnapshot();
  });

  it('content is text', () => {
    const html = render(<PageHeaderWrapper content="just so so" />).html();
    expect(html).toMatchSnapshot();
  });
});
