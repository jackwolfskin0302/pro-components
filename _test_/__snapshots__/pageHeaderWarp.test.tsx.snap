// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BasicLayout base use 1`] = `"<div class=\\"ant-pro-page-header-wrap-page-header-warp\\"><div class=\\"ant-pro-grid-content\\"><div class=\\"ant-page-header\\"><div class=\\"ant-page-header-title-view\\"><span class=\\"ant-page-header-title-view-title\\"><h4 class=\\"ant-typography\\" style=\\"margin:0\\"></h4></span></div></div></div></div>"`;

exports[`BasicLayout content is text 1`] = `"<div class=\\"ant-pro-page-header-wrap-page-header-warp\\"><div class=\\"ant-pro-grid-content\\"><div class=\\"ant-page-header\\"><div class=\\"ant-page-header-title-view\\"><span class=\\"ant-page-header-title-view-title\\"><h4 class=\\"ant-typography\\" style=\\"margin:0\\"></h4></span></div><div class=\\"ant-page-header-content-view\\"><div class=\\"ant-pro-page-header-wrap-detail\\"><div class=\\"ant-pro-page-header-wrap-main\\"><div class=\\"ant-pro-page-header-wrap-row\\"><div class=\\"ant-pro-page-header-wrap-content\\">just so so</div></div></div></div></div></div></div></div>"`;
